src/Fable.Transforms/Rust/Fable2Rust.fs

```diff
@@ -1777,13 +1908,52 @@ module Util =
-        let makeArm pat targetIndex boundValues =
+        let namesForIndex evalType evalName caseIndex = //todo refactor with below
+            match evalType with
+            | Fable.DeclaredType(entRef, genArgs) ->
+                let ent = com.GetEntity(entRef)
+                if ent.IsFSharpUnion then
+                    let unionCase = ent.UnionCases |> List.item caseIndex
+                    match evalName with
+                    | Some idName ->
+                        unionCase.UnionCaseFields
+                        |> Seq.mapi (fun i _field ->
+                            $"{idName}_{caseIndex}_{i}", idName, _field.FieldType
+                        )
+                    | _ -> Seq.empty
+                else Seq.empty
+            | _ -> Seq.empty
+            |> Seq.toList
+        let makeArm pat targetIndex boundValues (extraVals: (string * string * Fable.Type) list)=
             let attrs = []
             let guard = None // TODO:
-            let idents, bodyExpr = targets |> List.item targetIndex // TODO:
-            let vars = idents |> List.map (fun (ident: Fable.Ident) -> ident.Name)
+            let idents, (bodyExpr: Fable.Expr) = targets |> List.item targetIndex // TODO:
+            let vars = idents |> List.map (fun (id: Fable.Ident) -> id.Name)
             // TODO: vars, boundValues
-            let body = com.TransformAsExpr(ctx, bodyExpr)
+            let body =
+                //com.TransformAsExpr(ctx, bodyExpr)
+                let useages = UseageTracking.calcIdentUseages bodyExpr
+                let symbolsAndNames =
+                    let fromIdents =
+                        idents
+                        |> List.map(fun id ->
+                            id.Name, {  IsRef = true
+                                        IsMutable = id.IsMutable
+                                        IsRefCountWrapped = shouldBeRefCountWrapped com id.Type
+                                        MultipleUseages = UseageTracking.hasMultipleUseages id.Name useages })
+                    let fromExtra =
+                        extraVals
+                        |> List.map (fun (name, friendlyName, t) ->
+                            friendlyName, { IsRef = true
+                                            IsMutable = false
+                                            IsRefCountWrapped = shouldBeRefCountWrapped com t
+                                            MultipleUseages = UseageTracking.hasMultipleUseages friendlyName useages })
+                    fromIdents @ fromExtra
+                let scopedSymbolsNext =
+                    (ctx.ScopedSymbols, symbolsAndNames)
+                    ||> List.fold (fun acc (name, arg) -> acc |> Map.add name arg)
+                let ctx = { ctx with ScopedSymbols = scopedSymbolsNext; Typegen = { ctx.Typegen with TakingOwnership = true } }
+                transformLeaveContextByValue com ctx (bodyExpr.Type) None bodyExpr
             mkArm attrs pat guard body
```

Comment:
@alexswan10k Why is the arm body transformation so different than others bodies, can this be unified with function and lambda body processing? This is just for my information, perhaps I'm missing the context (no pun intended).

Reply:
Agree, this can almost definitely be unified and has turned into a mess. I was desperately trying to get it to work with the union tests (which all broke with value types after passing by ref), and in order to do it I needed to thread through the ref status of the bound variable, but binding does not seem to be implemented. This is almost certainly not right. I was hoping to just be able to enumerate the bound variables and push them into the context object, but that didn't seem to be possible.

I am happy to give this another pass tomorrow if you like, just not sure where to start with getting the bindings working tbh. I spent a bit of time debugging but it wasn't clear what was missing.

Reply:
Thanks, this is basic infrastructure so it will probably pay of to get it right. Let me know if you want me to merge as is, so you can iterate on it afterwards. Perhaps a small example of the problem with bindings, just for illustration.

Reply:
Sure, whatever works for you. Only merge it if It isn't making your life harder. I am happy to hack away on it for a bit longer otherwise. I agree though, this needs to be right.

The offending test was this one:
```
            pub enum MyUnion { Case1(i32), Case2(i32), Case3(i32), }
            #[test]
            pub fn Union_case_matching_works() {
                let x: Rc<crate::Fable::Tests::Union::MyUnion> =
                    Rc::from(crate::Fable::Tests::Union::MyUnion::Case1(5));
                let actual: i32 =
                    match x.as_ref() {
                        crate::Fable::Tests::Union::MyUnion::Case2(x_1_0) =>
                        *x_1_0,
                        crate::Fable::Tests::Union::MyUnion::Case3(x_2_0) =>
                        *x_2_0,
                        crate::Fable::Tests::Union::MyUnion::Case1(x_0_0) =>
                        *x_0_0, // so before this was coming through as  x_0_0, which is wrong
                    };
                assert_eq!(actual, 5)
            }
```

I could not work out where the bindings x_1_0 were coming from, and then after some debugging i realised they were being generated. Without telling the scoped expression that "x" is a reference type, it does not understand that it has to dereference it before binding to the declared let var.